### PR TITLE
CloudStack fails to migrate VM with volume when there are datadisks attatched

### PR DESCRIPTION
**Pull request by GabrielBrascher**
### Description

This PR fixes issue #5407. In summary, volumes that should not be migrated are still being mapped in `VirtualMachineManagerImpl.createStoragePoolMappingsForVolumes`.

To give some context: method `createStoragePoolMappingsForVolumes` is used only when migrating a VM with its volume(s), therefore via the API call [migrateVirtualMachineWithVolume](https://cloudstack.apache.org/api/apidocs-4.15/apis/migrateVirtualMachineWithVolume.html).

In such cases, there are two options: 
A. User does not provide volumes, just the VM ID, and Host ID
B. User provides VM ID, Host ID, and a volume map, e.g. `migrateto[0].volume=71f43cd6-69b0-4d3b-9fbc-67f50963d60b&migrateto[0].pool=a382f181-3d2b-4413-b92d-b8931befa7e1&migrateto[1].volume=88de0173-55c0-4c1c-a269-83d0279eeedf&migrateto[1].pool=95d6e97c-6766-4d67-9a30-c449c15011d1`.

However, regardless of case A or B, `createStoragePoolMappingsForVolumes` iterates all the VM's volumes and adds to be migrated. Therefore, this causes exceptions when migrating VMs with data disk attached whenever the data-disk (shared) should not be migrated.

From my point of view, it does not make sense to add volumes that the user has not mapped to be migrated unless the VM is migrated to **(I)** different cluster or **(II)** the volume scope is LOCAL. Both of these cases are already covered by the if statement previous to the removed line in this PR.

```
if (ScopeType.HOST.equals(currentPool.getScope()) || isStorageCrossClusterMigration(plan.getClusterId(), currentPool)) {
                createVolumeToStoragePoolMappingIfPossible(profile, plan, volumeToPoolObjectMap, volume, currentPool);
}
```

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [X] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Migrate VM with Root disk in local storage.

Tests have been done via the UI as well as the API. Also, it considered on both cases:
A. User does not provide volumes, just the VM ID, and HostID;
B. User provides VM ID, and HostID, and a volume map, e.g. `migrateto[0].volume=71f43cd6-69b0-4d3b-9fbc-67f50963d60b&migrateto[0].pool=a382f181-3d2b-4413-b92d-b8931befa7e1&migrateto[1].volume=88de0173-55c0-4c1c-a269-83d0279eeedf&migrateto[1].pool=95d6e97c-6766-4d67-9a30-c449c15011d1`.

In both cases, the volume map has been done accordingly with the expected behavior.

VMs that fail to be migrated, have been successfully migrated and only the correct volume was migrated; data disks remained attached to the VM, at their storage pool, and in "Ready" state.

### CODE REVIEW

**Comment by GabrielBrascher** [comment]
This has been tested with KVM.

Pinging some of the experts in managed storages (@mike-tutkowski @slavkap) to check if there are any cases I am missing where such volume mapping is correct.

Also, I would appreciate feedback regarding different hypervisors, such as VMware and Xen (@rhtyd @shwstppr @nvazquez @DaanHoogland).

**Review by nvazquez — COMMENTED** [review]
I think this would break at least Vmware offline migrations (https://github.com/apache/cloudstack/blob/main/engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java#L1314) cc. @rhtyd  @shwstppr @weizhouapache @sureshanaparti @davidjumani  - any input on this?

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 1]
```diff
@@ -2932,8 +2932,6 @@ protected void createStoragePoolMappingsForVolumes(VirtualMachineProfile profile
             executeManagedStorageChecksWhenTargetStoragePoolNotProvided(targetHost, currentPool, volume);
             if (ScopeType.HOST.equals(currentPool.getScope()) || isStorageCrossClusterMigration(plan.getClusterId(), currentPool)) {
                 createVolumeToStoragePoolMappingIfPossible(profile, plan, volumeToPoolObjectMap, volume, currentPool);
-            } else {
-                volumeToPoolObjectMap.put(volume, currentPool);
```
Comment (shwstppr) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2936]: @GabrielBrascher as you have seen `allVolumes` is a confusing name here. It is only the volumes that are not mapped using API. I've not tested but VMware might expect mappings for each of the volumes of the VM.
I feel instead of preventing adding a map here, we should look into the code that does migration and skip volumes that are already on the destination pool. For KVM this could be https://github.com/apache/cloudstack/blob/main/engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java#L1978-L1980

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
Tested offline migration on VMware env and **the volume for which mapping was not provided in API call got detached from the VM in vCenter**.
- Migrated a stopped with a ROOT and a DATA disk within the cluster.
- Both volumes were on same primary store before migration - **ps1**
- Migrated with `migrateVirtualMachineWithVolume` API by passing a mapping for ROOT volume for primary store **ps1-1**

**Comment by GabrielBrascher** [comment]
Thanks for reviewing and testing @shwstppr @nvazquez!

I will check different approaches as this could impact VMware migrations.
When another approach is ready for review I will ping you.

**Comment by GabrielBrascher** [comment]
@shwstppr @nvazquez PR is ready for review.

**Comment by GabrielBrascher** [comment]
I've found some issues when having Local + NFS attached, with Local + RBD data disk all works fine.

I will check the workflow with the NFS tests feedback.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by GabrielBrascher** [comment]
I am running a couple more tests. However, considering that the 4.16 cut will be soon, I might reduce the scope to fix only RBD (datadisk) + Local (Root). And then block migrations with NFS (datadisk) + Local (Root).

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java [hunk 2]
```diff
@@ -83,26 +82,51 @@
     @Override
     protected StrategyPriority internalCanHandle(Map<VolumeInfo, DataStore> volumeMap, Host srcHost, Host destHost) {
         if (super.internalCanHandle(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
-            if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
-                Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();
+            if (isKvmNfsStorageMigration(volumeMap, srcHost, destHost) || isLocalStorageMigration(volumeMap)) {
```
Comment (GutoVeronezi) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java:85]: We could join these `if` statements.

**Comment by sureshanaparti** [comment]
@GabrielBrascher I've tested VM migration with volumes (ROOT disk on local storage + DATA disk on shared storage [NFS]), to another host within the same cluster. Migration failed with/without volumes mapping. Did I missed anything? Let me know if I've to test with any specific steps (and configuration).

**Without volumes mapping =>**
```
(cloudcmk) > migrate virtualmachinewithvolume virtualmachineid=455dc09d-7b05-42ad-89fb-3b0bcb9dc01e hostid=bf2abca7-0b54-4bdd-a3dc-0056ec706e18 
{
  "accountid": "4f9b2858-1d1d-11ec-81eb-1e0079000453",
  "cmd": "org.apache.cloudstack.api.command.admin.vm.MigrateVirtualMachineWithVolumeCmd",
  "completed": "2021-09-24T11:02:43+0000",
  "created": "2021-09-24T11:02:40+0000",
  "jobid": "21f63e7d-967e-443d-8cbc-72322fb977a7",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 530,
    "errortext": "Failed to migrated vm VM instance {id: \"3\", name: \"i-2-3-VM\", uuid: \"455dc09d-7b05-42ad-89fb-3b0bcb9dc01e\", type=\"User\"} along with its volumes. "
  },
  "jobresultcode": 530,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "4f9cfd7a-1d1d-11ec-81eb-1e0079000453"
}
🙈 Error: async API failed for job 21f63e7d-967e-443d-8cbc-72322fb977a7
```

**With volumes mapping =>**
(cloudcmk) > migrate virtualmachinewithvolume virtualmachineid=455dc09d-7b05-42ad-89fb-3b0bcb9dc01e hostid=bf2abca7-0b54-4bdd-a3dc-0056ec706e18 migrateto[0].volume=898ee4b0-445a-4091-a402-0d52546f08c1 migrateto[0].pool=b67d9027-afcb-4a73-b24c-b1dc06a8d539 migrateto[1].volume=0d9efebe-c7ed-4030-92f9-33d4dfa50c87 migrateto[1].pool=3f039d92-52c0-3b68-88c5-8b27ea8e1fd2
{
  "accountid": "4f9b2858-1d1d-11ec-81eb-1e0079000453",
  "cmd": "org.apache.cloudstack.api.command.admin.vm.MigrateVirtualMachineWithVolumeCmd",
  "completed": "2021-09-24T11:12:41+0000",
  "created": "2021-09-24T11:12:38+0000",
  "jobid": "5badf4d5-c2f9-40e9-b859-c1c082dfc1d6",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 530,
    "errortext": "Failed to migrated vm VM instance {id: \"3\", name: \"i-2-3-VM\", uuid: \"455dc09d-7b05-42ad-89fb-3b0bcb9dc01e\", type=\"User\"} along with its volumes. "
  },
  "jobresultcode": 530,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "4f9cfd7a-1d1d-11ec-81eb-1e0079000453"
}
🙈 Error: async API failed for job 5badf4d5-c2f9-40e9-b859-c1c082dfc1d6

**Comment by sureshanaparti** [comment]
> @GabrielBrascher I've tested VM migration with volumes (ROOT disk on local storage + DATA disk on shared storage [NFS]), to another host within the same cluster. Migration failed with/without volumes mapping. Did I missed anything? Let me know if I've to test with any specific steps (and configuration).
> 
> **Without volumes mapping =>**
> 
> ```
> (cloudcmk) > migrate virtualmachinewithvolume virtualmachineid=455dc09d-7b05-42ad-89fb-3b0bcb9dc01e hostid=bf2abca7-0b54-4bdd-a3dc-0056ec706e18 
> {
>   "accountid": "4f9b2858-1d1d-11ec-81eb-1e0079000453",
>   "cmd": "org.apache.cloudstack.api.command.admin.vm.MigrateVirtualMachineWithVolumeCmd",
>   "completed": "2021-09-24T11:02:43+0000",
>   "created": "2021-09-24T11:02:40+0000",
>   "jobid": "21f63e7d-967e-443d-8cbc-72322fb977a7",
>   "jobprocstatus": 0,
>   "jobresult": {
>     "errorcode": 530,
>     "errortext": "Failed to migrated vm VM instance {id: \"3\", name: \"i-2-3-VM\", uuid: \"455dc09d-7b05-42ad-89fb-3b0bcb9dc01e\", type=\"User\"} along with its volumes. "
>   },
>   "jobresultcode": 530,
>   "jobresulttype": "object",
>   "jobstatus": 2,
>   "userid": "4f9cfd7a-1d1d-11ec-81eb-1e0079000453"
> }
> 🙈 Error: async API failed for job 21f63e7d-967e-443d-8cbc-72322fb977a7
> ```
> 
> **With volumes mapping =>**
> 
> ```
> (cloudcmk) > migrate virtualmachinewithvolume virtualmachineid=455dc09d-7b05-42ad-89fb-3b0bcb9dc01e hostid=bf2abca7-0b54-4bdd-a3dc-0056ec706e18 migrateto[0].volume=898ee4b0-445a-4091-a402-0d52546f08c1 migrateto[0].pool=b67d9027-afcb-4a73-b24c-b1dc06a8d539 migrateto[1].volume=0d9efebe-c7ed-4030-92f9-33d4dfa50c87 migrateto[1].pool=3f039d92-52c0-3b68-88c5-8b27ea8e1fd2
> {
>   "accountid": "4f9b2858-1d1d-11ec-81eb-1e0079000453",
>   "cmd": "org.apache.cloudstack.api.command.admin.vm.MigrateVirtualMachineWithVolumeCmd",
>   "completed": "2021-09-24T11:12:41+0000",
>   "created": "2021-09-24T11:12:38+0000",
>   "jobid": "5badf4d5-c2f9-40e9-b859-c1c082dfc1d6",
>   "jobprocstatus": 0,
>   "jobresult": {
>     "errorcode": 530,
>     "errortext": "Failed to migrated vm VM instance {id: \"3\", name: \"i-2-3-VM\", uuid: \"455dc09d-7b05-42ad-89fb-3b0bcb9dc01e\", type=\"User\"} along with its volumes. "
>   },
>   "jobresultcode": 530,
>   "jobresulttype": "object",
>   "jobstatus": 2,
>   "userid": "4f9cfd7a-1d1d-11ec-81eb-1e0079000453"
> }
> 🙈 Error: async API failed for job 5badf4d5-c2f9-40e9-b859-c1c082dfc1d6
> ```

Will re-test with qemu ev pkgs and update here.

**Comment by sureshanaparti** [comment]
Tested with qemu ev pkgs, still migration failed with/without volumes mapping for the VM with ROOT disk on local storage + DATA disk on shared storage. Able to migrate VM with ROOT disk on shared storage + DATA disk on shared storage.

**Comment by rhtyd** [comment]
@GabrielBrascher can you test with ev packages, use mbx with CentOS7 if that helps.

**Comment by GabrielBrascher** [comment]
@sureshanaparti thanks for testing.

With Root Local + Datadisk (NFS) it is failing for me as well. However, it did not fail with just Local.

I decided to revert some changes and change the approach, in a few minutes I will update here and explain the reasons behind the new approach to tackle this issue.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by GabrielBrascher** [comment]
@rhtyd @nvazquez @shwstppr @GutoVeronezi thanks for the review. I am adding a few contexts and conclusions on how this implementation has evolved. I did a few changes on past Friday and have tested it since then [see [this comment](https://github.com/apache/cloudstack/pull/5410#issuecomment-926646375)].

#### Recap and adding context on this PR

##### There were two options to fix:
1. Before calling the copy command -- map only the volumes that are going to be migrated which are either just the root disk, or root disk (Local) + data disks (shared) that the user has explicitly added to be migrated in the API call.
2. At the `KvmNonManagedStorageDataMotionStrategy` -- this change would impact on the "canHandle" and add a few changes in the workflow in order to "ignore" some of the mapped volumes; which require more work to ensure the volumes should indeed be ignored.

##### Considerations for each option

**Option 1:** it is interesting as so far works like a charm and it is simple to implement; however, it requires specific logic for KVM in the "generic" class for migrating VM.
**Option 2:** makes sense to narrow it to the KVM  strategy; however, at that point, all VMs are mapped as migrating in the DB and it does not make much sense to have such volumes in the migrating state. The implementation also was a lot more complex to be done.


#### Tests and conclusion
I was successful with the Local + RBD migrations. looks like a great addition to 4.16 in this scenario.
However, Local + NFS still not working on my tests.

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java [hunk 3]
```diff
@@ -82,16 +82,13 @@
...
-            if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
-                Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();
-
-                for (VolumeInfo volumeInfo : volumeInfoSet) {
-                    StoragePoolVO storagePoolVO = _storagePoolDao.findById(volumeInfo.getPoolId());
-
-                    if (!supportStoragePoolType(storagePoolVO.getPoolType())) {
-                        return StrategyPriority.CANT_HANDLE;
-                    }
+        if (super.internalCanHandle(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE
```
Comment (GutoVeronezi) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java:85]: We could invert this `if`, to reduce indentation, and, maybe, add some logs.

**Comment by sureshanaparti** [comment]
@GabrielBrascher I've tested VM migration with volumes. Results below.

- **VM with Local ROOT disk**

(i) Without volumes mapping:
`migrate virtualmachinewithvolume virtualmachineid=4c5cf633-482d-499f-8d39-970fa1f62efc hostid=d19ac1e9-990d-404d-b60c-a7b75cf7836c`

=> Failed with error: "_Failed to migrated vm VM instance {id: \"7\", name: \"i-2-7-VM\", uuid: \"4c5cf633-482d-499f-8d39-970fa1f62efc\", type=\"User\"} along with its volumes._ "

(ii) With volumes mapping (Local storage on src host : Local storage on dest host):
`migrate virtualmachinewithvolume virtualmachineid=4c5cf633-482d-499f-8d39-970fa1f62efc hostid=d19ac1e9-990d-404d-b60c-a7b75cf7836c migrateto[0].volume=703abf09-99f7-48b8-9a47-55a467452a60 migrateto[0].pool=cae2c100-cfd8-47f1-aed5-3442b289ec85`

=> Failed with error: "_Failed to migrated vm VM instance {id: \"7\", name: \"i-2-7-VM\", uuid: \"4c5cf633-482d-499f-8d39-970fa1f62efc\", type=\"User\"} along with its volumes._ "

- **VM with Local ROOT disk + Shared (NFS) DATA disk**

(i) Without volumes mapping:
`migrate virtualmachinewithvolume virtualmachineid=3a879829-466a-451a-bb2d-a2897349a9d9 hostid=d19ac1e9-990d-404d-b60c-a7b75cf7836`

=> Passed

(ii) With ROOT disk volume mapping (Local storage on src host : Local storage on dest host):
`migrate virtualmachinewithvolume virtualmachineid=3a879829-466a-451a-bb2d-a2897349a9d9 hostid=d19ac1e9-990d-404d-b60c-a7b75cf7836c  migrateto[0].volume=c01302c0-41e4-4fbe-876e-13c3493ad8f6 migrateto[0].pool=cae2c100-cfd8-47f1-aed5-3442b289ec85`

=> Passed

(iii) With ROOT & DATA disk volume mapping (Local storage on src host : Local storage on dest host + NFS storage1 : NFS storage2):
`migrate virtualmachinewithvolume virtualmachineid=3a879829-466a-451a-bb2d-a2897349a9d9 hostid=d19ac1e9-990d-404d-b60c-a7b75cf7836c  migrateto[0].volume=c01302c0-41e4-4fbe-876e-13c3493ad8f6 migrateto[0].pool=cae2c100-cfd8-47f1-aed5-3442b289ec85 migrateto[1].volume=9f2f170a-6650-4750-bd9c-76fca2ae70cb migrateto[1].pool=df3aba77-9ff6-3c91-9062-f2d79686672e`

=> Failed with error: "_Failed to migrated vm VM instance {id: \"5\", name: \"i-2-5-VM\", uuid: \"3a879829-466a-451a-bb2d-a2897349a9d9\", type=\"User\"} along with its volumes._ "

- **VM with Shared (NFS) ROOT disk**

(i) With ROOT disk volume mapping:
`migrate virtualmachinewithvolume virtualmachineid=b56072e6-88d1-4c13-a6e7-b9ab76e54b82 migrateto[0].volume=bde7a8ee-85ed-4ed5-bfd4-4c2de38fba04 migrateto[0].pool=df3aba77-9ff6-3c91-9062-f2d79686672e hostid=999460c3-07e7-4eae-867a-c3747ce36589`

=> Failed with error: "_Failed to migrated vm VM instance {id: \"3\", name: \"i-2-3-VM\", uuid: \"b56072e6-88d1-4c13-a6e7-b9ab76e54b82\", type=\"User\"} along with its volumes._ "

- **VM with Shared (NFS) ROOT disk + Shared (NFS) DATA disk**

(i) With ROOT & DATA volumes mapping:
`migrate virtualmachinewithvolume virtualmachineid=2d94bc0e-9de1-4d0f-ae67-bf69286a2fdf hostid=999460c3-07e7-4eae-867a-c3747ce36589 migrateto[0].volume=6f99fb94-b4d9-462f-8e26-227f3d33f946 migrateto[0].pool=df3aba77-9ff6-3c91-9062-f2d79686672e migrateto[1].volume=2b1ec21e-d02a-441d-ab3e-ca831eb4c6a6 migrateto[1].pool=df3aba77-9ff6-3c91-9062-f2d79686672e`

=> Failed with error: "_Failed to migrated vm VM instance {id: \"6\", name: \"i-2-6-VM\", uuid: \"2d94bc0e-9de1-4d0f-ae67-bf69286a2fdf\", type=\"User\"} along with its volumes._ "

**Comment by sureshanaparti** [comment]
@GabrielBrascher I noticed the most of the migration with volumes fails, not sure if these are regression. Let me know if I have to check any specific cases.

**Comment by nvazquez** [comment]
@GabrielBrascher @sureshanaparti is this PR ready?

**Comment by sureshanaparti** [comment]
> @GabrielBrascher @sureshanaparti is this PR ready?

Not yet @nvazquez , Observed some failures in manual tests, waiting for the confirmation from @GabrielBrascher.

**Comment by GabrielBrascher** [comment]
@sureshanaparti sorry for the delay, I have been debugging and running some tests:

I am also pushing a commit (tested recently on it as well) which I am addressing a review from @GutoVeronezi.
These are the output of my recent tests:

#### VM With local Root disk
**(i) Without volumes mapping:**
migrate virtualmachinewithvolume virtualmachineid=1f83b074-2d25-4eb1-9e10-bf26ccb71e50 hostid=a19dcabd-53b6-44e4-9f57-2fa70d876ef8
=> Success

**(ii) With ROOT volume mapping:**
migrate virtualmachinewithvolume virtualmachineid=1f83b074-2d25-4eb1-9e10-bf26ccb71e50 hostid=448a0207-975b-41e1-af07-0c6e55dae0b3 migrateto[0].volume=e6e28756-0beb-4f75-8e6e-9ba5cc50b983 migrateto[0].pool=887fe175-7422-4e3a-a9ed-392932afa631
=> Success

#### VM With local Root disk + Shared (RBD) DATA disk
**(i) Without volumes mapping:**
migrate virtualmachinewithvolume virtualmachineid=1f83b074-2d25-4eb1-9e10-bf26ccb71e50 hostid=448a0207-975b-41e1-af07-0c6e55dae0b
=> Success

**(ii) With ROOT volume mapping:**
migrate virtualmachinewithvolume virtualmachineid=0f6c2e2d-7960-4715-b0c9-8634ccc36548 hostid= migrateto[0].volume=e6e28756-0beb-4f75-8e6e-9ba5cc50b983 migrateto[0].pool=2bfa2e18-d76f-46db-8605-4aafb1438030
=> Success


##### VM With local Root disk + Shared (NFS) DATA disk
**(i) Without volumes mapping:**
migrate virtualmachinewithvolume virtualmachineid=1f83b074-2d25-4eb1-9e10-bf26ccb71e50 hostid=448a0207-975b-41e1-af07-0c6e55dae0b
=> Failed with error:`"Copy volume(s) to storage(s) [{volume: "13433", from: "23", to:"26"}] and VM to host [{vm: "13452", from: "24", to:"46"}] failed in StorageSystemDataMotionStrategy.copyAsync. Error message: [Exception during migrate: org.libvirt.LibvirtException: internal error: unable to execute QEMU command 'nbd-server-add': Block node is read-only]."`


In conclusion, all tests worked fine for me, except by the one with NFS datadisk (which seem to work with you).
I assume that the exception might be due to configurations in the NFS pool or the Libvirt, due to "LibvirtException: internal error: unable to execute QEMU command 'nbd-server-add': Block node is read-only."

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java [hunk 4]
```diff
@@ -82,21 +82,19 @@
      */
     @Override
     protected StrategyPriority internalCanHandle(Map<VolumeInfo, DataStore> volumeMap, Host srcHost, Host destHost) {
-        if (super.internalCanHandle(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
-            if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
-                Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();
-
-                for (VolumeInfo volumeInfo : volumeInfoSet) {
-                    StoragePoolVO storagePoolVO = _storagePoolDao.findById(volumeInfo.getPoolId());
+        if (super.internalCanHandle(volumeMap, srcHost, destHost) != StrategyPriority.CANT_HANDLE
```
Comment (GabrielBrascher) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java:85]: @GutoVeronezi if logic has been inverted.

- From
```
if (super.internalCanHandle == StrategyPriority.CANT_HANDLE) {
   if (canHandleKVMNonManagedLiveNFSStorageMigration == StrategyPriority.CANT_HANDLE) {
      ...
   }
}
return StrategyPriority.CANT_HANDLE;
```

OR

```
if (super.internalCanHandle == StrategyPriority.CANT_HANDLE)
   && (canHandleKVMNonManagedLiveNFSStorageMigration == StrategyPriority.CANT_HANDLE) {
   ...
} 
return StrategyPriority.CANT_HANDLE;
```

- To
```
if (super.internalCanHandle != StrategyPriority.CANT_HANDLE
                || canHandleKVMNonManagedLiveNFSStorageMigration != StrategyPriority.CANT_HANDLE) {
            return StrategyPriority.CANT_HANDLE;
}
return StrategyPriority.HYPERVISOR;

```

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test keepEnv

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java [hunk 5]
```diff
@@ -2947,12 +2947,25 @@ protected void createStoragePoolMappingsForVolumes(VirtualMachineProfile profile
...
+            } else if (shouldMapVolume(profile, volume, currentPool)){
                 volumeToPoolObjectMap.put(volume, currentPool);
             }
         }
     }
 
+    /**
+     * Returns true if it should map the volume for a storage pool to migrate.
+     * <br><br>
+     * Some context: VMware migration workflow requires all volumes to be mapped (even if volume stays on its current pool);
```
Comment (sureshanaparti) [inline engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2959]: The current condition checks for not KVM. Do we have to check for VMware only, to map all volumes?

Comment (GabrielBrascher) [reply engine/orchestration/src/main/java/com/cloud/vm/VirtualMachineManagerImpl.java:2959]: @sureshanaparti thanks again for reviewing.

My main idea on isolating this on KVM is due to the fact that I was able to reproduce/debug this issue only on KVM. Additionally, I am taking into account the `KvmNonManagedStorageDataMotionStrategy` migration process.

Therefore, considering the review and tests from @shwstppr with the fact that I am seeing this in KVM I changed the mapping just for the KVM.
The question would be, does this issue is also happening on XenServer? If anyone can check it on a XenServer env it would be great.

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by sureshanaparti** [comment]
@blueorangutan test matrix

**Comment by sureshanaparti** [comment]
> @sureshanaparti @vladimirpetrov @borisstoyanov @nvazquez do we need any further manual testing before merging this?

@rhtyd yes, requires some manual tests for VMware migration cases.
